src/docs/arch-usage-patterns.dox

```diff
@@ -0,0 +1,223 @@
+or pointer indicating the slot is initialized prior to reading the contents of the object held in
+the slot.
+
+@subsubsection slotted-arrays-correctness Correctness of allowing reading while allowing writes
+Additional writes to the array can occur concurrently, and as a copy of \c slot_count is taken by
+readers prior to their traversal they may miss objects being initialized concurrently. The reader
+algorithm must take care to ensure that once it has entered the loop new slots being used in the
+array will not impact the outcome of the loop as it is not guaranteed to see those.
+
+@subsection growable-arrays Growable arrays
+A growable array is one where the value of \c Z is not fixed and the user can increase the array
+size while the array is in use. Here we will discuss the additional semantics required to support
+a growable array.
+
+__Note:__ @ref generations handle the life cycle, creation and freeing of the growable array and
+that will not be covered here.
+
+@subsubsection growable-arrays-writing Writing
+If a new slot is required in the array, but all slots are in use and \c slot_count is equal to the
+size \c Z of the array then the array must grow. The variable \c slot_count has additional ordering
+defined when swapping in the array. The array pointer swap must happen before \c slot_count is
+incremented. This requires a @ref rel-marked.
+
+@subsubsection growable-arrays-reading Reading
+Readers must respect the writer side ordering. The \c slot_count must be read prior to reading the
+array pointer. This requires an @ref acq-marked. If this ordering is not respected the reader could
+read an old array pointer but a new \c slot_count and encounter an array index out of bounds
+exception.
+
+@subsubsection growable-arrays-further-details Further details
+The fixed size version of this pattern is used to maintain the @ref arch-connection's sessions
+array. The dynamically growing version of this pattern is used to maintain the hazard pointer array
+for a @ref arch-session.
+
+@section write-while-stable Write While stable (WwS)
+The Write while Stable (WwS) pattern performs a write to a memory location \c W, but only considers
+that write to be successful if the value of a *separate* memory location \c S has not changed by the
+time the write to \c W has become visible. As a result when the \c WwS pattern succeeds, the next
+time the value of \c S changes all threads can already see the update to \c W.
+
+In this pattern the write to \c W always takes place, however the writing thread will not consider
+the update to be valid until it has verified that \c S is unchanged. The writing thread is
+responsible for either retrying or rolling back the write to \c W on a failure. Threads reading from
+\c W must be able to handle or ignore these interim "failed" writes to \c W, and must perform a
+read-acquire on \c S before reading \c W.
+
+<img src="WwS.png" alt="" style="transform: scale(0.75,0.75);" />
```

Comment:
nit: This image is rather large in the generated docs. 
I tried changing scale to (0.5, 0.5) but it still looked ugly so happy to leave as is if we can't find a quick fix.

Reply:
Yeah, I'm not sure how well the scale config works but I'll leave it as is for now.